File: Quorum/Library/Standard/Plugins/javascript/Libraries/Game/Graphics/Painter3D.js
function plugins_quorum_Libraries_Game_Graphics_Painter3D_() 
{
    var context = new plugins_quorum_Libraries_Game_Graphics_RenderContext_();
    var shaderProvider = new plugins_quorum_Libraries_Game_Graphics_ShaderProvider_();
    var environment;
    var isRendering = false;
    var skybox = null;
    var skyboxShader = new plugins_quorum_Libraries_Game_Graphics_SkyboxShader_();
    
    var quorumBatch;
    var renderables;
    var camera;
    
    /*
    Used to detect if we've rendered the Skybox yet.
    We render the skybox after all opaque objects and before the first
    blended (i.e. transparent) objects.
    */
    var shouldRenderSkybox = false;
    
    this.Initialize$quorum_Libraries_Game_Graphics_Painter3D$quorum_Libraries_Containers_Array = function(batch, render) 
    {
        quorumBatch = batch;
        renderables = render;
    };
    
    this.IsRendering = function() 
    {
        return isRendering;
    };
    
    this.SetCamera$quorum_Libraries_Game_Graphics_Camera = function(cam) 
    {
        if (isRendering && renderables.GetSize() > 0)
            this.Flush();
        
        camera = cam;
    };
    
    this.GetCamera = function() 
    {
        return camera;
    };
    
    this.SetEnvironment$quorum_Libraries_Game_Graphics_Environment = function(environ) 
    {
        environment = environ;
    };
    
    this.GetEnvironment = function() 
    {
        return environment;
    };

    this.SetSkybox$quorum_Libraries_Game_Graphics_Skybox = function(box)
    {
        skybox = box;
    };
    
    this.GetSkybox = function()
    {
        return skybox;
    };

    this.Begin = function() 
    {
        if (isRendering)
        {
            var exceptionInstance_ = new quorum_Libraries_Language_Errors_Error_();
            exceptionInstance_.SetErrorMessage$quorum_text("The Painter3D is already rendering! Call End() before calling Begin() again.");
            throw exceptionInstance_;
        }
        if (camera === null || camera === undefined)
        {
            var exceptionInstance_ = new quorum_Libraries_Language_Errors_Error_();
            exceptionInstance_.SetErrorMessage$quorum_text("The Painter3D must have a camera set before calling Begin().");
            throw exceptionInstance_;
        }
        
        context.Begin();
        isRendering = true;
        
        shouldRenderSkybox = !(skybox === null || skybox === undefined);
    };
    
    this.End = function() 
    {
        this.Flush();
        
        // Render the Skybox if we haven't already.
        if (shouldRenderSkybox)
        {
            if (!skybox.plugin_.SidesRequested())
            {
                var exceptionInstance_ = new quorum_Libraries_Language_Errors_Error_();
                exceptionInstance_.SetErrorMessage$quorum_text("I can't render the skybox because it wasn't fully loaded. Make sure all six sides are loaded before trying to use it.");
                throw exceptionInstance_;
            }
            else if (skybox.IsLoaded())
                skyboxShader.Render(skybox, camera);
        }
        context.End();
        isRendering = false;
    };
    
    this.Flush = function()
    {
        renderables.Sort();
        var currentShader = null;
        var renderable;
        var renderPlugin;
        
        if (shouldRenderSkybox && !skybox.plugin_.SidesRequested())
        {
            var exceptionInstance_ = new quorum_Libraries_Language_Errors_Error_();
            exceptionInstance_.SetErrorMessage$quorum_text("I can't render the skybox because it wasn't fully loaded. Make sure all six sides are loaded before trying to use it.");
            throw exceptionInstance_;
        }
        
        for (var i = 0; i < renderables.GetSize(); i++)
        {
            renderable = renderables.Get$quorum_integer(i);
            renderPlugin = renderable.plugin_;
            
            /*
            When we find our first blended object, render the skybox first.
            The Skybox should be in the background before any transparent
            objects are drawn, so they can blend the skybox into the rendered
            pixel fragments.
            */
            if (shouldRenderSkybox && renderable.UsesBlending())
            {
                shouldRenderSkybox = false;
                if (currentShader !== null && currentShader !== undefined)
                    currentShader.End();
                currentShader = null;
                skyboxShader.Render(skybox, camera);
            }
            
            if (currentShader !== renderPlugin.shader)
            {
                if (currentShader !== null && currentShader !== undefined)
                    currentShader.End();
                currentShader = renderPlugin.shader;
                currentShader.Begin(camera, context);
            }
            currentShader.Render(renderable);
        }
        if (currentShader !== null && currentShader !== undefined)
            currentShader.End();
        
        renderables.Empty();
    };
    
    this.RenderNative$quorum_Libraries_Game_Graphics_Renderable = function(renderable) 
    {
        renderable.Set_Libraries_Game_Graphics_Renderable__environment_(environment);
        var renderPlugin = renderable.plugin_;
        renderPlugin.camera = camera;
        renderPlugin.shader = shaderProvider.GetShader(renderable);
    };
}

File: Plugins/GamePlugin/JavaGame/src/plugins/quorum/Libraries/Game/Graphics/Painter3D.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package plugins.quorum.Libraries.Game.Graphics;

import plugins.quorum.Libraries.Game.Graphics.Shaders.SkyboxShader;
import plugins.quorum.Libraries.Game.GameRuntimeError;
import plugins.quorum.Libraries.Game.GameStateManager;
import quorum.Libraries.Game.Graphics.Painter3D_;
import quorum.Libraries.Game.Graphics.Camera_;
import quorum.Libraries.Game.Graphics.Camera;
import quorum.Libraries.Game.Graphics.Renderable_;
import quorum.Libraries.Containers.Array_;
import quorum.Libraries.Game.Graphics.Environment_;
import quorum.Libraries.Game.Graphics.Skybox_;

/**
 *
 * @author alleew
 */
public class Painter3D 
{
    public java.lang.Object me_ = null;
    
    protected RenderContext context = new RenderContext();
    protected ShaderProvider shaderProvider = new ShaderProvider();
    protected Environment_ environment = null;
    protected Skybox_ skybox = null;
    protected SkyboxShader skyboxShader = new SkyboxShader();
    
    private boolean isRendering = false;
    
    // A convenient way to access the quorum side.
    private quorum.Libraries.Game.Graphics.Painter3D quorumBatch;
    
    private Array_ renderables;
    
    /*
    Used to detect if we've rendered the Skybox yet.
    We render the skybox after all opaque objects and before the first
    blended (i.e. transparent) objects.
    */
    private boolean shouldRenderSkybox = false;
    
    public void Initialize(Painter3D_ batch, Array_ array)
    {
//        quorumBatch = (quorum.Libraries.Game.Graphics.Painter3D)batch;
        renderables = array;
    }
    
    public boolean IsRendering()
    {
        return isRendering;
    }
    
    public void SetCamera(Camera_ cam)
    {
        if (isRendering && renderables.GetSize() > 0)
            Flush();
        
//        camera = cam;
    }
    
    public Camera_ GetCamera()
    {
//        return camera;
        return null;
    }
    
    public void SetEnvironment(Environment_ environ)
    {
        environment = environ;
    }
    
    public Environment_ GetEnvironment()
    {
        return environment;
    }

    public void SetSkybox(Skybox_ box)
    {
        skybox = box;
    }
    
    public Skybox_ GetSkybox()
    {
        return skybox;
    }
    
    public void Begin()
    {
        if (isRendering)
            throw new GameRuntimeError("The Painter3D is already rendering! Call End() before calling Begin() again.");
//        if (camera == null)
//            throw new GameRuntimeError("The Painter3D must have a camera set before calling Begin().");
        
        context.Begin();
        isRendering = true;
        
        shouldRenderSkybox = (skybox != null);
    }
    
    public void End()
    {
        Flush();
        
        // Render the Skybox if we haven't already.
        if (shouldRenderSkybox)
        {   
            if (skybox.IsLoaded() == false)
                throw new GameRuntimeError("I can't render the skybox because it wasn't fully loaded. Make sure all six sides are loaded before trying to use it.");
//            skyboxShader.Render(skybox, camera);
        }
        context.End();
        isRendering = false;
    }
    
    public void Flush()
    {
        renderables.Sort();
        Shader currentShader = null;
        
        if (shouldRenderSkybox && skybox.IsLoaded() == false)
            throw new GameRuntimeError("I can't render the skybox because it wasn't fully loaded. Make sure all six sides are loaded before trying to use it.");
        
        for (int i = 0; i < renderables.GetSize(); i++)
        {
            Renderable_ renderable = (Renderable_)renderables.Get(i);
//            Renderable renderPlugin = ((quorum.Libraries.Game.Graphics.Renderable)renderable).plugin_;
            
            /*
            When we find our first blended object, render the skybox first.
            The Skybox should be in the background before any transparent
            objects are drawn, so they can blend the skybox into the rendered
            pixel fragments.
            */
            if (shouldRenderSkybox && renderable.UsesBlending())
            {
                shouldRenderSkybox = false;
                if (currentShader != null)
                    currentShader.End();
                currentShader = null;
//                skyboxShader.Render(skybox, camera);
            }
            
//            if (currentShader != renderPlugin.shader) 
            {
                if (currentShader != null)
                    currentShader.End();
//                currentShader = renderPlugin.shader;
//                currentShader.Begin(camera, context);
            }
            currentShader.Render(renderable);
        }
        if (currentShader != null) {
            currentShader.End();
        }
        
        if(!renderables.IsEmpty()) {
            renderables.Empty(false);
        }
    }
    
//    public void RenderNative(Renderable_ renderable)
//    {
//        renderable.Set_Libraries_Game_Graphics_Renderable__environment_(environment);
//        Renderable renderPlugin = ((quorum.Libraries.Game.Graphics.Renderable)renderable).plugin_;
//        renderPlugin.camera = camera;
//        renderPlugin.shader = shaderProvider.GetShader(renderable);
//    }
}
